### ISSUE TITLE
Confused about Access Control mechanism

### ISSUE THREAD

**Issue by ezekieltsui** [CESNET/netopeer2#754]
# Confused about Access Control mechanism
Hi, Michal,
I have some confusion about NACM and need your help.

**Background :** 

As user `netconf`, which was automatically created by Dockerfile, I want to send a `<rpc>` message( edit-config ) with `netopeer2-cli`. I use `user-rpc` command and enter the following `<rpc>` message:

```xml
<edit-config xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
        <target>
            <running/>
        </target>
        <config>
                <cont xmlns="urn:examples">
	    <l>1080</l>
               </cont>	
        </config>
</edit-config>
```

But I always get an ERROR as below :

```
ERROR
	type:     protocol
	tag:      access-denied
	severity: error
	path:     /examples:cont/l
	message:  Access to the data model "examples" is denied because "netconf" NACM authorization failed.
```

**Preparation :**

environment : 

​	docker ( I pulled the image  2 months ago ), seems no important commit in this 2 months.

So, a system user `netconf` was created, which belongs to group `sudo`.

Then,

I installed `examples` module :

```
# sysrepoctl -i sysrepo/examples/examples.yang
```

changed `examples` module permisson and owner:

```
# sysrepoctl -c examples -p 666 -g sudo
```

check if I successfully install `examples` module with :

```
# sysrepoctl -l
```

and i found the flag of `examples` module is `I`, which means Installed. and permisson si `666`, owner is `root:sudo`, everything seems to be as expected.

**Test :**

I did 3 tests in three situations :

<u>Note:</u> `<rpc>` message is the same as above. 

Situation 1:

```
# netopeer2-cli
> connect --ssh --host localhost --port 830 --login netconf
> user-rpc		// send edit-config <rpc> as user netconf 
```

result : got NACM ERROR

Situation 2:

```
# netopeer2-cli
> connect --ssh --host localhost --port 830 --login root
> disconnect	// just disconnect, send nothing as user root
> connect --ssh --host localhost --port 830 --login netconf
> user-rpc		// send edit-config <rpc> as user netconf
```

result : got NACM ERROR

Situation 3:

```
# netopeer2-cli
> connect --ssh --host localhost --port 830 --login root
> user-rpc	// send edit-config <rpc> as user root
> disconnect
> connect --ssh --host localhost --port 830 --login netconf
> user-rpc  // send edit-config <rpc> as user netconf
```

result : OK

emmm, is this as expected ? Or do u think it is a bug ? I notice u said it's odd in [issue #747](https://github.com/CESNET/netopeer2/issues/747) from netopeer2 project, an issue seems similar to mine.

==============================================================================

Later, in [issue #1766](https://github.com/sysrepo/sysrepo/issues/1766) from sysrepo project, I found u metioned that there 2 access control policies. ( file system for `sysrepo` and NACM for `NETCONF`) 

Above tests I have change the file system permisson of `examples` module to `666`, but I still blocked by NACM. So I think I also need to configure NACM configuration. Means need to configure `ietf-netconf-acm` module ? 

So I import an NACM configuration to `ietf-netconf-acm` module with this :

```
# sysrepocfg --import=ietf-netconf-acm.xml -d startup -m ietf-netconf-acm
```

``` xml
<!-- ietf-netconf-acm.xml -->
<nacm xmlns="urn:ietf:params:xml:ns:yang:ietf-netconf-acm">
  <groups>
    <group>
      <name>sudo</name>
      <user-name>netconf</user-name>
    </group>
  </groups>
  <rule-list>
    <name>test-rule-list</name>
    <group>sudo</group>
    <rule>
      <name>test-rule-name</name>
      <module-name>examples</module-name>
      <path>/</path>
      <access-operations>*</access-operations>
      <action>permit</action>
      <comment>test-comment</comment>
    </rule>
  </rule-list>
</nacm>
```

I can get the configuration I imported before with :

```
# sysrepocfg -X -d startup -m ietf-netconf-acm
```

But still got ERROR same as the beginning.

Next, I change the import command to (change the parameter of `-d` from `startup` to `running`) :

```
# sysrepocfg --import=ietf-netconf-acm.xml -d running -m ietf-netconf-acm
```

It work, I can send a send a `<rpc>` message( edit-config ) as user `netconf`.

But after I replace the `ietf-netconf-acm` configuration with empty configuration, it still work, user `netconf` didnt blocked by NACM again. That's strange. Once I successfully send a `<rpc>` message, no matter what I do to the configuration, it always work, unless I create a new docker contatiner.

**Qestions :**

To sum up, there are questions I want to ask:

​	a) After I successfully send the `<rpc>` message with `netconf` account, how can I withdraw the permission ?

​	b) How to get `ietf-netconf-acm` module' s configuration using `netopeer2-cli ` ?

​	c) No matter what I did, the flag of `examples` module always be `I`, how to change this flag to `i` ?

​	d) Why should I set the `-d` parameter to `running` instead of `startup`? I think I misunderstanding something, I will have a further reading first.
 
===================================================

BTW, the online document of `libnetconf2` is gone, show 404 ERROR.

**Comment by michalvasko** [comment]
Hi,
firstly, you should understand when exactly *sysrepo* file system access control is used and when *netopeer2* NACM. I think it is well summarized in *netopeer2* [README](https://github.com/CESNET/Netopeer2/#nacm).

But to explain a bit, whenever you are directly calling *sysrepo* API functions in an application or using `sysrepoctl`/`sysrepocfg`, *sysrepo* access control is used. And when you are connected to *netopeer2* with `netopeer2-cli`, then NACM is used and *sysrepo* access control ignored. Now I can try to answer.

a) This is a wrong understanding of how it works and should be just an illusion. NACM can at times work in a complex and confusing way so you would need to specify exact use-case, what is the result and what result you would expect. Also, it is still possible there is a bug causing some invalid behavior.

b) NACM [blocks reading](https://github.com/CESNET/netopeer2/blob/master/modules/ietf-netconf-acm%402018-02-14.yang#L214) its configuration by default for all the users so you can read it only if you have explicit access to it or you are using the recovery user (`root` by default).

c) I am not sure you understand what the flag means. Assuming you do, simply uninstall the module and if another module needs it imported, it will do so automatically.

d) It depends what datastore you want to write to, there is no inherent reason why you should write into `running` instead of `startup`.

What *libnetconf2* document are you referring to? Is there a link referenced that is not working anymore?

**Comment by EzekielTsui** [comment]
Hi, thanks for your reply.

Now I understand the relationship between NACM and sysrepo access control.

a ) 

<u>use-case</u> :

```
root@b3a70104919e:/# netopeer2-cli 
> connect --ssh --host localhost --port 830 --login netconf
Interactive SSH Authentication
Type your password:
Password: 
> user-rpc 
ERROR
	type:     protocol
	tag:      access-denied
	severity: error
	path:     /examples:cont/l
	message:  Access to the data model "examples" is denied because "netconf" NACM authorization failed.

> disconnect 
> connect --ssh --host localhost --port 830 --login root
Interactive SSH Authentication
Type your password:
Password: 
> user-rpc
OK
> disconnect 
> connect --ssh --host localhost --port 830 --login netconf
Interactive SSH Authentication
Type your password:
Password: 
> user-rpc 
OK
> 
```

I don’t understand why the reply I get when I send `<rpc>` for the **second** time is "OK", during this period, I did not modify any NACM configuration. Shouldn't it be blocked by NACM again ?

( Yes, it's an illusion, should not happen in the actual production environment, maybe we can ignore it. )

b ) and c ) I understood, thx.

d ) At first I thought that restarting `netopeer2-server` would copy the `startup` datastore to the `running` datastore, but in fact I need to reboot my device. Clear now.

Thanks a lot.

===================================================================================

The original link of `libnetconf2` online document is https://netopeer.liberouter.org/doc/libnetconf2/devel/index.html. I can find other project's documents( `libyang` and `sysrepo` ) except for `libnetconf2`.

**Comment by michalvasko** [comment]
a) What exactly is the RPC you are sending?

The documentation should work again.

**Comment by EzekielTsui** [comment]
a ) Like I said above, I have installed the `examples` module. And the content of the RPC is :
```xml
<edit-config xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
        <target>
            <running/>
        </target>
        <config>
                <cont xmlns="urn:examples">
	    <l>1080</l>
               </cont>	
        </config>
</edit-config>
```


Yes, I can access the document now.

**Comment by EzekielTsui** [comment]
typo correction :
I don’t understand why the reply I get when I send <rpc> for the **third** time is "OK", during this period, I did not modify any NACM configuration. Shouldn't it be blocked by NACM again ?

**Comment by michalvasko** [comment]
Right, it is working correctly since the RPC was `edit-config`. When applying a configuration from `edit-config`, the effective operation for each node is first determined. In the first case, `l` was supposed to be created, I assume, which was blocked. In the second case it was not blocked, so `l` was created. In the third case, `l` already existed so the effective operation was `none`, which [requires no access rights](https://tools.ietf.org/html/rfc8341#section-3.2.5) so it passed without any error.

**Comment by ezekieltsui** [comment]
Great! Thanks for your explanation.
